File: app/src/main/java/com/shuoye/video/fragment/home/HomeFragment.java
package com.shuoye.video.fragment.home;

import android.os.Bundle;
import android.os.Handler;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.viewpager2.adapter.FragmentStateAdapter;

import com.google.android.material.tabs.TabLayoutMediator;
import com.shuoye.video.R;
import com.shuoye.video.databinding.FragmentHomeBinding;
import com.shuoye.video.utils.Utils;
import com.youth.banner.indicator.CircleIndicator;

import org.jetbrains.annotations.NotNull;

import java.util.Date;
import java.util.Objects;


/**
 * 首页
 *
 * @author shuoye
 */
public class HomeFragment extends Fragment {
    FragmentHomeBinding binding;
    Handler.Callback callback;
    NavController navController;
    HomeViewModel model;
    private static final String TAG = "Liu_shuoye";
    private static final String[] TABS = Utils.getArray(R.array.week_array);

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentHomeBinding.inflate(inflater);
        model = new ViewModelProvider(this).get(HomeViewModel.class);
        //获取导航控制器
        navController = Navigation.findNavController(requireActivity(), R.id.fragmentContainerView);
        //TODO 设置点击跳转
        //binding.animeList.setOnClickListener(v -> navController.navigate(R.id.action_homeFragment_to_animeListFragment));
        callback = msg -> {
            if (msg.what == R.id.setBannerAdapter) {
                //设置轮播图适配器
                setBannerAdapter();
            } else if (msg.what == R.id.setTimeLine) {
                //设置新番时间表
                setTimeLine();
            }
            return true;
        };

        if (model.getCallback() == null) {
            model.setCallback(callback);
        } else {
            //设置轮播图适配器
            setBannerAdapter();
            //更新番剧列表
            setTimeLine();
        }
        return binding.getRoot();
    }

    /**
     * 设置轮播图适配器
     */
    void setBannerAdapter() {
        binding.banner.setAdapter(new ImageAdapter(model.getDataBeans()))
                .addBannerLifecycleObserver(this)
                .setIndicator(new CircleIndicator(getContext()))
                .setLoopTime(4000);
        binding.banner.setOnBannerListener((data, position) -> {
            Log.d(TAG, data.getTitle());
            Bundle bundle = new Bundle();
            bundle.putInt("id", data.getAID());
            //TODO 点击跳转到详情
//            navController.navigate(R.id.action_homeFragment_to_animeInfoFragment, bundle);
        });//设置点击事件

    }

    /**
     * 设置新番时间表
     */
    void setTimeLine() {
        binding.viewPager2.setAdapter(new FragmentStateAdapter(this) {
            @NotNull
            @Override
            public Fragment createFragment(int position) {
                return new TimeLineFragment(model.getTimeLines().get(position));
            }

            @Override
            public int getItemCount() {
                return 7;
            }
        });

        new TabLayoutMediator(binding.tabLayout, binding.viewPager2, (tab, position) -> tab.setText(TABS[position])).attach();

        long time = System.currentTimeMillis();
        Date date = new Date(time);
        Objects.requireNonNull(binding.tabLayout.getTabAt(Utils.getWeekOfDate(date))).select();

    }
}

File: app/src/main/java/com/shuoye/video/ui/home/HomeFragment.kt
package com.shuoye.video.ui.home

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.findNavController
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import com.shuoye.video.database.pojo.Banner
import com.shuoye.video.databinding.FragmentHometBinding
import com.shuoye.video.ui.home.adapters.ImageBannerAdapter
import com.shuoye.video.utils.Utils
import com.youth.banner.indicator.CircleIndicator
import com.youth.banner.listener.OnBannerListener
import dagger.hilt.android.AndroidEntryPoint
import java.util.*


@AndroidEntryPoint
class HomeFragment : Fragment() {

    private val model: HomeViewModel by viewModels()
    private lateinit var binding: FragmentHometBinding

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentHometBinding.inflate(inflater, container, false)
        context ?: return binding.root
        init()
        return binding.root
    }

    private fun init() {
        // 初始化轮播图
        setBannerAdapter(binding)
        //初始化新番时间表
        initTimeLine(binding)

        //初始化点击事件
        binding.update.setOnClickListener { view ->
            val action = HomeFragmentDirections.actionHomeFragmentToUpdateFragment()
            view.findNavController().navigate(action)
        }
        binding.recommend.setOnClickListener { view ->
            val action = HomeFragmentDirections.actionHomeFragmentToRecommendFragment()
            view.findNavController().navigate(action)
        }
        binding.search.setOnClickListener { view ->
            val action = HomeFragmentDirections.actionHomeFragmentToSearchFragment()
            view.findNavController().navigate(action)
        }
    }

    /**
     * 设置轮播图适配器
     */
    private fun setBannerAdapter(binding: FragmentHometBinding) {
        model.getBannerLiveData().observe(viewLifecycleOwner, {
            binding.banner
                .addBannerLifecycleObserver(viewLifecycleOwner)
                .setIndicator(CircleIndicator(context))
                .setLoopTime(4000)
                .setAdapter(ImageBannerAdapter(it.data as MutableList<Banner>?))
                .setOnBannerListener(OnBannerListener { banner: Banner, _: Int ->
                    val action = HomeFragmentDirections.actionHomeFragmentToAnimeInfoFragment(
                        banner.id
                    )
                    binding.root.findNavController().navigate(action)
                })
        })
    }

    /**
     * 初始化番剧更新时间表
     */
    private fun initTimeLine(binding: FragmentHometBinding) {
        binding.viewPager2.adapter = object : FragmentStateAdapter(this) {
            override fun createFragment(position: Int): Fragment {
                return TimeLineFragment((position + 1) % 7)
            }

            override fun getItemCount(): Int {
                return 7
            }
        }
        TabLayoutMediator(
            binding.tabLayout,
            binding.viewPager2
        ) { tab: TabLayout.Tab, position: Int ->
            when (position) {
                0 -> tab.text = "一"
                1 -> tab.text = "二"
                2 -> tab.text = "三"
                3 -> tab.text = "四"
                4 -> tab.text = "五"
                5 -> tab.text = "六"
                6 -> tab.text = "日"
            }
        }.attach()
        val time = System.currentTimeMillis()
        val date = Date(time)
        binding.tabLayout.getTabAt(Utils.getWeekOfDate(date))?.select()

    }
}

